### PR TITLE
CrossEntropyLoss triggers floating point exception

### LINKED ISSUE #73165: `CrossEntropyLoss` triggers floating point exception
### 🐛 Describe the bug

`CrossEntropyLoss` triggers floating point exception

```python
import torch
results = dict()
input_1 = torch.rand([5, 0], dtype=torch.float32)
input_2 = torch.rand([5, 0], dtype=torch.float32)

torch.nn.CrossEntropyLoss()(input_1, input_2)
# floating point exception
```

### Versions

pytorch: 1.10.1

cc @ezyang @gchanan @zou3519 @albanD @mruberry @jbschlosser @walterddr @kshitij12345

### PR DESCRIPTION
**Pull request by mikeb127**
Adds cast to float to fix floating point exception

Fixes #73165

### CODE REVIEW

**Comment by mikeb127** [comment]
Reviewers - Apologies for what may be a stupid question, but I'm new to contributing to this repo....

The tests above that are failing seem to have nothing to do with what I changed at all. Do I need to address these before you guys will review/accept?

File: aten/src/ATen/native/LossNLL.cpp [hunk 1]
```diff
@@ -491,7 +491,7 @@ Tensor cross_entropy_loss_prob_target(
 
     switch (reduction) {
       case Reduction::Mean:
-        return -(input * target * weight_).sum() / (input.numel() / input.size(1));
+        return -(input * target * weight_).sum() / ((float) input.numel() / (float) input.size(1));
```
Comment (jbschlosser) [inline aten/src/ATen/native/LossNLL.cpp:494]: Please use C++-style casts (i.e. `static_cast<float>(...)`) over C-style.

**Review by jbschlosser — CHANGES_REQUESTED** [review]
Thanks for the fix! Sorry for the review delay - I was out last week.

I'm not convinced this solves the problem of 0 in the denominator. Please add a test in `test/test_nn.py` ensuring that the case referenced in the issue causes no issue after the fix:
```python
input_1 = torch.rand([5, 0], dtype=torch.float32)
input_2 = torch.rand([5, 0], dtype=torch.float32)
torch.nn.CrossEntropyLoss()(input_1, input_2)
```

File: aten/src/ATen/native/LossNLL.cpp (hunk 1 above)
Comment (mikeb127) [reply aten/src/ATen/native/LossNLL.cpp]: Changes made and tests added. Apologies for the ridiculous number of commits, I have manage to absolutely butcher a simple change.

Comment (jbschlosser) [reply aten/src/ATen/native/LossNLL.cpp]: No worries at all! No change is simple in this codebase tbh :)

Comment (mikeb127) [reply aten/src/ATen/native/LossNLL.cpp]: Thanks :). Anything else you need from me for this one? Do you want me to try and rebase the PR to get rid of these unrelated errors? Or are you happy to review as is?

Comment (mikeb127) [reply aten/src/ATen/native/LossNLL.cpp]: Converted back to draft to fix failing tests

Comment (mikeb127) [reply aten/src/ATen/native/LossNLL.cpp]: All tests now passing. Let me know what you think

**Review by jbschlosser — APPROVED** [review]
LGTM thanks!

**Pull request closed without merging**